### PR TITLE
#324: Added check that abstract imports are in abstract modules

### LINKED ISSUE #324: module annotation causes invalid C# code
Consider `CompileRefinement.dfy`:

```
abstract module Sig {
    class Foo {
        const v: int
        constructor () {}
        method hi()
    }
}

module Impl1 refines Sig {
    class Foo {
        const v := 42
        constructor () {}
        method hi() {
            print "Hello!\n";
        }
    }
}

module Program {
    import opened Impl1: Sig

    method Main() {
        var f := new Foo();
        // as expected, we get an assertion violation here because
        // the value of v is not specified in Sig
        // assert f.v == 42;
        f.hi();
    }
}
```

If I try to compile this file to C#, an invalid C# file is created:

```
$ dafny CompileRefinement.dfy /compile:3 /spillTargetCode:2 /out:./CompileRefinementOut
Dafny 2.3.0.10506

Dafny program verifier finished with 1 verified, 0 errors
Compiled program written to CompileRefinementOut.cs
Errors compiling program
c:\Users\kggruett\AppData\Local\Temp\wyqfk3ev.0.cs(1553,7) : error CS0246: The type or namespace name '_0_Sig' could not be found (are you missing a using directive or an assembly reference?)

c:\Users\kggruett\AppData\Local\Temp\wyqfk3ev.0.cs(1554,22) : error CS0246: The type or namespace name '_0_Sig' could not be found (are you missing a using directive or an assembly reference?)
```

This is a bug: The Dafny-to-C# compiler should never spit out invalid C# files. Either it should spit out a valid C# file, or produce an error saying why it is not capable of doing so.

Interestingly, if I comment out the `: Sig` annotation in the import, everything works as expected:

```
$ dafny CompileRefinement_noSigAnnotation.dfy /compile:3 /spillTargetCode:2 /out:./CompileRefinement_noSigAnnotationOut
Dafny 2.3.0.10506

Dafny program verifier finished with 1 verified, 0 errors
Compiled program written to CompileRefinement_noSigAnnotationOut.cs
Running...

Hello!
```

The relevant diff between the two created `.cs` files is:

```
1553,1554c1553,1554
<       _0_Sig.Foo _13_f;
<       var _nw0 = new _0_Sig.Foo();
---
>       _1_Impl1_Compile.Foo _12_f;
>       var _nw0 = new _1_Impl1_Compile.Foo();
```

No code is emitted for the definitions in `Sig`, because it's an abstract module, but yet, the constructor `new _0_Sig.Foo();` is called. The correct thing for the compiler to do would be to somehow bypass the `: Sig` annotation and use `new _1_Impl1_Compile.Foo();` in both cases.
Or, if that's too hard, it should emit an error saying `module annotations are not supported for compilation`.

### PR DESCRIPTION
**Pull request by davidcok**
Fixes #324.

### CODE REVIEW

**Comment by RustanLeino** [comment]
I like the description "abstract import". I like it! I think we've used more confusing words before. "Abstract import" more directly says that the import is not complete.

File: Source/Dafny/Resolver.cs [hunk 1]
```diff
@@ -1435,6 +1435,11 @@ public class ModuleBindings
           dependencies.AddEdge(decl, d);
           var subbindings = bindings.SubBindings(d.Name);
           ProcessDependencies(d, subbindings ?? bindings, dependencies);
+          if (!m.IsAbstract && d is ModuleFacadeDecl && ((ModuleFacadeDecl) d).Root != null) {
+            reporter.Error(MessageSource.Resolver, d.tok, 
+              "The abstract import named {0} (using :) may only occur within an abstract module declaration",
```
Comment (RustanLeino) [inline Source/Dafny/Resolver.cs:1440]: What do you think of changing "occur within" to "be used in"?

File: Test/git-issues/git-issue-324.dfy [hunk 2]
```diff
@@ -0,0 +1,37 @@
+// RUN: %dafny /compile:0 "%s" > "%t"
+// RUN: %dafny /noVerify /compile:4 /compileTarget:cs "%s" >> "%t"
+// RUN: %dafny /noVerify /compile:4 /compileTarget:js "%s" >> "%t"
+// RUN: %dafny /noVerify /compile:4 /compileTarget:go "%s" >> "%t"
+// RUN: %dafny /noVerify /compile:4 /compileTarget:java "%s" >> "%t"
+// RUN: %diff "%s.expect" "%t"
+
+abstract module Sig {
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-324.dfy:8]: Use 2 spaces as indentation throughout

File: Test/git-issues/git-issue-324.dfy [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+module Program {
+    import opened Impl1: Sig
+
+    method Main() {
+        var f := new Foo();
+        // as expected, we get an assertion violation here because
+        // the value of v is not specified in Sig
+        // assert f.v == 42;
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-324.dfy:33]: This is old, right? Delete it.

File: Source/Dafny/Resolver.cs (hunk 1 above)
Comment (davidcok) [reply Source/Dafny/Resolver.cs:1440]: Agree

File: Test/git-issues/git-issue-324.dfy (hunk 2 above)
Comment (davidcok) [reply Test/git-issues/git-issue-324.dfy:8]: Removed

File: Test/git-issues/git-issue-324.dfy (hunk 3 above)
Comment (davidcok) [reply Test/git-issues/git-issue-324.dfy]: Removed

**Pull request merged**